test/e2e/upgrade_test.go

```diff
@@ -217,153 +227,150 @@ var _ = Describe("BMO Upgrade", func() {
 		}
 	})
 
-	It("Should upgrade BMO to latest version", func() {
-		if e2eConfig.GetVariable("UPGRADE_DEPLOY_IRONIC") != "false" {
-			// Install Ironic
-			ironicKustomization := e2eConfig.GetVariable("IRONIC_KUSTOMIZATION")
-			By("Installing Ironic on the upgrade cluster")
-			err := BuildAndApplyKustomization(ctx, &BuildAndApplyKustomizationInput{
-				Kustomization:       ironicKustomization,
-				ClusterProxy:        upgradeClusterProxy,
-				WaitForDeployment:   true,
-				WatchDeploymentLogs: true,
-				DeploymentName:      "ironic",
-				DeploymentNamespace: bmoIronicNamespace,
-				LogPath:             filepath.Join(artifactFolder, "logs", fmt.Sprintf("%s-%s", bmoIronicNamespace, specName)),
-				WaitIntervals:       e2eConfig.GetIntervals("default", "wait-deployment"),
+	for i := range bmoFromKustomizations {
```

Comment:
This is not a great pattern. We end up using indices all over the place and it becomes hard to understand and read the code.
Could you instead break out the test spec into a separate function that takes a struct with config as input? Then we just call that function for each test we want to run. ([Example](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/1cb5d5329f967e90ed891386f598cb266e8daa13/test/e2e/suites/e2e/clusterctl_upgrade_test.go#L33-L100))

Reply:
Yes, thanks for the idea. I kinda like it to be in a loop though, but having a function is also nice. Extra useful since we will be able to use it for ironic upgrade as well.

Reply:
Thanks Lennart. And sorry for not getting back to you sooner.

I think using the same function for both of upgrade cases is possible, and maybe even preferable, as most of the steps are identical. That's up for discussion though, but in this PR I will focus on making the BMO upgrade first.

Regarding the loop, using DescribeTable is a good idea, and I think I will take that into use, but my reason for looping was that I don't want to pre-define the number of test cases, but rather based on the number of kustomizations provided by the user. Do you happen to have a good idea how to solve that, or maybe my expectation was just to unrealistic?

Reply:
I think I've come up with a good solution. Could you please have a look?